### PR TITLE
feat: set up Git remotes from factory url

### PR DESCRIPTION
**Pull request by dkwon17**
Signed-off-by: dkwon17 <dakwon@redhat.com>

### What does this PR do?
Adds a `remotes` url parameters for the factory url to set up remotes in the devworkspace (in the `spec.template.projects` section)

**IMPORTANT! A few notes about this PR:**
* If the repo has no devfile, the default devfile is used and  a project will be added with the remotes configured
* If the repo's devfile has no `projects` field, a project will be added with the remotes configured
* If the repo's devfile has a `projects` field with one `git` project, that project will have the remotes configured
* If the repo's devfile has a `projects` field with more than one `git` project, remotes are not configured since it is unclear what project to configure remotes for.
  * [error message](https://user-images.githubusercontent.com/83611742/208253900-a590a655-2ebf-43af-b9e3-67c3f0908cb2.png)
  * If this is a problem, we can change the syntax used for `remotes` url parameter to specify a project name (`spec.template.projects[*].name`) for the remotes to specifically configure

* It is not yet possible to set what exact branch to checkout, as this PR only introduces the ability to configure remotes. The default branch for the first remote will be checked out when starting the workspace.

### What issues does this PR fix or reference?
Part of https://issues.redhat.com/browse/CRW-3482 and https://github.com/eclipse/che/issues/21315

### Is it tested? How?

As explained in https://issues.redhat.com/browse/CRW-3482, remotes can be configured in the url as follows:
```
# 1. explicit name and URL
?remotes={{origin,https://github...},{upstream,https://github...},...}

# 2. URLs only, name is implicit: first is origin, second is upstream, subsequent are fork1, fork2 etc... 
?remotes={https://github...,https://github...,...}
```

Deploy Che and set the following in the CheCluster CR:
```
spec:
  components:
    dashboard:
      deployment:
        containers:
          - image: 'quay.io/dkwon17/che-dashboard:remotes'
            name: che-dashboard
```

To test, are some scenarios:

1. URLs only, name is implicit.
  Go to: `{CHE-HOST}/#https://github.com/eclipse-che/che-plugin-registry?remotes={https://github.com/dkwon17/che-plugin-registry,https://github.com/eclipse-che/che-plugin-registry,https://github.com/che-dockerfiles/che-plugin-registry}`
  And verify that the resulting devworkspace's yaml has the following remotes:
```
    projects:
      - git:
          checkoutFrom:
            remote: origin
          remotes:
            fork1: 'https://github.com/che-dockerfiles/che-plugin-registry'
            origin: 'https://github.com/dkwon17/che-plugin-registry'
            upstream: 'https://github.com/eclipse-che/che-plugin-registry'
        name: che-plugin-registry
```
2. Explicit name and URL.
  Go to: `{CHE-HOST}/#https://github.com/eclipse-che/che-plugin-registry?remotes={{origin,https://github.com/dkwon17/che-plugin-registry},{upstream,https://github.com/eclipse-che/che-plugin-registry}}`
  And verify that the resulting devworkspace's yaml has the following remotes:
```
    projects:
      - git:
          checkoutFrom:
            remote: origin
          remotes:
            origin: 'https://github.com/dkwon17/che-plugin-registry'
            upstream: 'https://github.com/eclipse-che/che-plugin-registry'
```
3. Explicit name and URL, but change the order of the remotes to verify that `checkoutFrom` has changed.
  Go to: `{CHE-HOST}/#https://github.com/eclipse-che/che-plugin-registry?remotes={{upstream,https://github.com/eclipse-che/che-plugin-registry},{origin,https://github.com/dkwon17/che-plugin-registry}}`
```
    projects:
      - git:
          checkoutFrom:
            remote: upstream
          remotes:
            origin: 'https://github.com/dkwon17/che-plugin-registry'
            upstream: 'https://github.com/eclipse-che/che-plugin-registry'
```
4. If the remotes cannot be parsed due to syntax errors, or if it is empty, this error appears:
image

### CODE REVIEW

File: packages/dashboard-frontend/src/containers/Loader/Factory/__tests__/getGitRemotes.spec.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+import { getGitRemotes, GitRemote } from '../getGitRemotes';
+
+describe('getGitRemotes()', () => {
```
Comment (dkwon17) [inline packages/dashboard-frontend/src/containers/Loader/Factory/__tests__/getGitRemotes.spec.ts:15]: If the code looks ok, I plan to add/fix more tests

File: packages/dashboard-frontend/src/services/workspace-client/devworkspace/__tests__/devWorkspaceClient.create.spec.ts [hunk 2]
```diff
@@ -103,6 +104,7 @@ describe('DevWorkspace client, create', () => {
         undefined,
         undefined,
         'eclipse/theia/next',
+        [],
```
Comment (dkwon17) [inline packages/dashboard-frontend/src/services/workspace-client/devworkspace/__tests__/devWorkspaceClient.create.spec.ts:107]: If the code looks ok, I plan to add more tests in this file

**Comment by l0rd** [comment]
@dkwon17 there is one scenario I am interested in: factory URL support for repositories that are not on GitHub, BB or GitLab (for example [sourcehut](https://sourcehut.org/), [gerrit](https://www.gerritcodereview.com/), [gogs](https://gogs.io/), [gitea](https://gitea.io/),[gitolite](https://gitolite.com/)). In these cases we are not able to retrieve the devfile so the workspace startup is failing. And that's ok.

But with the `remotes` parameter I am expecting that `https://<che-host>?remotes={<git-remote-url>}` would actually start a workspace and clone `<git-remote-url>`. For example `https://<che-host>?remotes={https://git.sr.ht/~geb/numen}`. The devfile, if it exist, would not be retrieved and applied but at list the factory URL can start the workspace. Could you please test if that's working?

File: packages/dashboard-frontend/src/services/workspace-client/devworkspace/__tests__/devWorkspaceClient.create.spec.ts (hunk 2 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/services/workspace-client/devworkspace/__tests__/devWorkspaceClient.create.spec.ts:107]: The code looks good.

File: packages/dashboard-frontend/src/containers/Loader/Factory/__tests__/getGitRemotes.spec.ts (hunk 1 above)
Comment (olexii4) [reply packages/dashboard-frontend/src/containers/Loader/Factory/__tests__/getGitRemotes.spec.ts:15]: I added feedback for 'sanitizeValue' https://github.com/eclipse-che/che-dashboard/pull/691/files#r1049678740

**Comment by olexii4** [comment]
@dkwon17 
BTW: you should fix the license in your code.

File: packages/dashboard-frontend/src/containers/Loader/Factory/getGitRemotes.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+function convertToArrayString(str: string): string {
+  return (
+    str
+      .replaceAll(' ', '')
+      .replaceAll('{', '[')
+      .replaceAll('}', ']')
+      /* eslint-disable no-useless-escape */
+      .replaceAll(/(\[([^\[]))/g, '["$2')
+      .replaceAll(/(([^\]])\])/g, '$2"]')
+      .replaceAll(/(([^\]]),([^\[]))/g, '$2","$3')
```
Comment (amisevsk) [inline packages/dashboard-frontend/src/containers/Loader/Factory/getGitRemotes.ts:67]: Consider including short comments to make parsing these ones a little easier (especially with the capture groups), e.g. `replace '[<string>' with '["<string>`

**Review by amisevsk — COMMENTED** [review]
Code-wise looks ok to me; haven't tested changes directly.

Regarding the comment in the PR description
> The remotes specified in the remotes url parameter will override the existing value for spec.template.projects[0]. 

This may be acceptable as default behavior (i.e. no error for devfiles with multiple projects). I'm not entirely sure how the dashboard handles the various cases (repo with devfile and no projects, repo with devfile and projects withing, repo with no devfile), but our focus is generally devfiles with one project, and that project should be the one the URL references. In that case it makes sense to just add remotes to that project.

There is an issue (that I can't find at the moment) for deprecating `projects` and replacing it with `project`, to host a single project also.

**Comment by l0rd** [comment]
[The deprecate `projects` issue](https://github.com/eclipse/che/issues/21188)

File: packages/dashboard-frontend/src/containers/Loader/Factory/getGitRemotes.ts (hunk 3 above)
Comment (dkwon17) [reply packages/dashboard-frontend/src/containers/Loader/Factory/getGitRemotes.ts]: Thank you, fixed

**Comment by dkwon17** [comment]
@olexii4 @akurinnoy I've updated the PR to configure project remotes in the devfile level instead of the devworkspace level, how does it look? I did this to make it a bit easier to facilitate https://github.com/eclipse-che/che-dashboard/pull/691#issuecomment-1352919656

@l0rd I made some updates to the PR, starting a workspace with:
```
https://<che-host>#/load-factory?remotes={https://git.sr.ht/~geb/numen}
```

does clone the repo specified in the `remotes`, and uses the default devfile. For now, the url needs the `#/load-factory` part.

<details><summary>Starting workspace</summary>

https://user-images.githubusercontent.com/83611742/208251535-09134f2d-5871-46cc-8148-e2f3c9fbda57.mov

</details>

<details><summary>An error sometimes appears, but doesn't affect workspace startup, workspace still starts, need to investigate</summary>

image

</details>

**Comment by dkwon17** [comment]
> The remotes specified in the remotes url parameter will override the existing value for spec.template.projects[0].

@amisevsk my comment there is a bit outdated now, I've updated the PR description for more details on specific cases on repos with/without devfiles and projects: https://github.com/eclipse-che/che-dashboard/pull/691#issue-1497536247

**Comment by dkwon17** [comment]
> <details><summary>An error sometimes appears, but doesn't affect workspace startup, workspace still starts, need to investigate</summary>

image

</details>


It seems like this is happening because the OAuth token refresh is being attempted, related issue: https://github.com/eclipse/che/issues/21896

**Review by ibuziuk — APPROVED** [review]

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/index.spec.tsx [hunk 4]
```diff
@@ -151,13 +155,224 @@ describe('Factory Loader container, step CREATE_WORKSPACE__APPLYING_DEVFILE', ()
...
+
+      factoryId = `${REMOTES_ATTR}=${remotesAttr}&` + factoryId;
+
+      await waitFor(() =>
+        expect(prepareDevfile).toHaveBeenCalledWith(devfile, factoryId, undefined, false),
+      );
+
+      expect(devfile.projects).not.toBe(undefined);
+      expect(devfile.projects!.length).toBe(1);
+      expect(devfile.projects![0]).toMatchObject({
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/index.spec.tsx:183]: Using the optional chaining operator is safer than the non-null assertion operator. Besides, you will not need to disable the corresponding `eslint` rule to make it happy.

```suggestion
      expect(devfile.projects?.length).toBe(1);
      expect(devfile.projects?.[0]).toMatchObject({
```

**Review by akurinnoy — APPROVED** [review]
The code looks good; there are only a few minor nitpicks.

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/getGitRemotes.spec.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+import { getGitRemotes, GitRemote, sanitizeValue } from '../getGitRemotes';
+
+describe('getGitRemotes functions', () => {
+  describe('getGitRemotes()', () => {
+    it('should return remotes when two values', () => {
```
[the remark this answers is not in the document]
Comment (dkwon17) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/getGitRemotes.spec.ts]: Oops, thank you,
I've updated it to be a bit more descriptive:
https://github.com/eclipse-che/che-dashboard/blob/274e8a5f90dd3cfce366db562eda47b4462ce322/packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/getGitRemotes.spec.ts#L45

File: packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/getGitRemotes.spec.ts [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    it('should return remote when two values', () => {
+      const input =
+        '{https://github.com/test1/che-dashboard, https://github.com/test2/che-dashboard}';
+      const expected: GitRemote[] = [
+        { name: 'origin', url: 'https://github.com/test1/che-dashboard' },
+        { name: 'upstream', url: 'https://github.com/test2/che-dashboard' },
+      ];
+      expect(getGitRemotes(input)).toMatchObject(expected);
+    });
+    it('should return remotes when two values', () => {
```
[the remark this answers is not in the document]
Comment (dkwon17) [reply packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/getGitRemotes.spec.ts]: Yes, you're right,
I've removed it and added a new test:

https://github.com/eclipse-che/che-dashboard/blob/274e8a5f90dd3cfce366db562eda47b4462ce322/packages/dashboard-frontend/src/containers/Loader/Factory/Steps/Apply/Devfile/__tests__/getGitRemotes.spec.ts#L75-L82

**Review by olexii4 — APPROVED** [review]

**Pull request merged**